### ISSUE TITLE
Major Rewrite (v2.x)

### ISSUE THREAD

**Issue by Drizin** [Drizin/DapperQueryBuilder#61]
I've been considering a major rewrite for a while, for many reasons:
- Current version is too coupled with Dapper, it's barely reusable
- Too difficult to customize or modify anything. Bad design decisions, I guess
- Merging SQL statements is just too ugly and the way we're doing [regex conversion](https://github.com/Drizin/DapperQueryBuilder/blob/master/src/DapperQueryBuilder/ParameterInfos.cs#L193) back and forth is [not efficient](https://github.com/Drizin/DapperQueryBuilder/issues/53). The idea is that we should use a standard format for the underlying string and only do the conversion (and give names to the numbered args) by the moment that we need to invoke Dapper.
- net6.0+ support InterpolatedStringHandlers which don't even require regex parsing

As a first draft, I've created this fresh project which is just a `FormattableString` implementation, but it's like a StringBuilder since it supports Append, Replace, Insert, etc: [https://github.com/Drizin/InterpolatedStrings](https://github.com/Drizin/InterpolatedStrings).

The next major release of DapperQueryBuilder will be based on this new library. (I guess major parts of CommandBuilder will be delegated to this other library).

Feel free to share your feedback.

**Comment by Drizin** [comment]
Sorry for spamming everyone - but there are so few contributors that I took the liberty to ping you all to get your feedback:

@jehhynes @frankyjquintero @terryaney @LarsStegman @erenguler @mullon

**Comment by terryaney** [comment]
Love re-writes when it cleans up/removes code and makes it more correct! :)  Couple of easy'ish questions

1. Is the idea then to re-write DapperQueryBuilder, just leveraging InterpolatedStrings? (either by you or contributors)
2. In InterpolatedStrings, as someone with zero multi framework support experience, how come the `*6` methods are required?  That can't just be detected automatically *inside* the method and use the appropriate language feature?  Surprised me when I saw that, but I'm sure I'm about to be educated ;)

**Comment by jehhynes** [comment]
Is there any reason you would want to use the pre-net6 version when using net6? If not, couldn't you do the following to avoid the strange `*6` methods?

```
    public class InterpolatedStringFactory
    {
#if NET6_0_OR_GREATER
        /// <summary>an
        /// Creates a new InterpolatedStringBuilder using an InterpolatedStringHandler (net6.0+) which can be a little faster than using regex.
        /// </summary>
        public virtual InterpolatedStringBuilder Create(InterpolatedStringAdapter value)
        {
            return value.InterpolatedStringBuilder;
        }

        /// <summary>
        /// Creates a new InterpolatedStringBuilder using an InterpolatedStringHandler (net6.0+) which can be a little faster than using regex.
        /// </summary>
        public virtual InterpolatedStringBuilder Create([InterpolatedStringHandlerArgument("options")] InterpolatedStringAdapter value, InterpolatedStringBuilderOptions options)
        {
            return value.InterpolatedStringBuilder;
        }
#else
        /// <summary>
        /// Creates a new InterpolatedStringBuilder using regular expression for parsing the FormattableString. 
        /// </summary>
        public virtual InterpolatedStringBuilder Create(FormattableString source)
        {
            var target = new InterpolatedStringBuilder();
            Parser.ParseAppend(source, target);
            return target;
        }
#endif
```

Also, you could easily support all the way back to netstandard1.4 (and thus support the .NET Framework back to 4.6.1). To do this:
1) You have an unneeded using statement to remove in `InterpolatedStringParser`: 
`using static System.Net.WebRequestMethods;`
2) And make the following change in `InterpolatedStringArgumentComparer`
```
        public int GetHashCode(
#if NETCOREAPP
            [DisallowNull] 
#endif
        InterpolatedStringArgument obj)
```

**Comment by Drizin** [comment]
> 1. Is the idea then to re-write DapperQueryBuilder, just leveraging InterpolatedStrings? (either by you or contributors)

Great question. I still don't have an answer.
InterpolatedStringBuilder provides some extensibility points but many of them are based on inheritance/overriding, and DapperQueryBuilder will need those overrides to do its magic (like automatically adding missing spaces, or automatically stripping single quotes around parameters). The bad part of using pure inheritance is that InterpolatedStringBuilder has a Fluent API (many methods return the same InterpolatedStringBuilder - the concrete type), and 
we would have to use method hiding (`new`) to change the types. And the terminology is just not intuitive (calling it Arguments instead of SqlParameters). 

So I think composition (where inheritance is only used in a nested class to do the overrides) is a better approach, and I already have a draft for that - I'm just not sure if this extra overhead (and complexity) is acceptable or not. 

The other possible approach is to just copy the whole source from InterpolatedStrings, and do the required changes. Probably breaking it into a base class (protected methods with different name) and InterpolatedSqlBuilder would inherit and expose the public methods. (still assuming that Fluent API will be based on a concrete type, and not on an interface).

> 2. In InterpolatedStrings, as someone with zero multi framework support experience, how come the `*6` methods are required?  That can't just be detected automatically _inside_ the method and use the appropriate language feature?  Surprised me when I saw that, but I'm sure I'm about to be educated ;)

That was just laziness on my part. I should probably just use the same names and for net6+ we should just prefer the new version (hide the regex one). I was a little afraid that the 2 different versions (the regex and the InterpolatedStringHandler) are not yet 100% identical behavior - I guess we just need some more tests. That's why it's still "beta". But the idea is that as soon as I confirm that both versions are working identically we should have a single method available. Sorry for the confusion.

**Comment by jehhynes** [comment]
> > Ok, PR created :-) All tests are passing. [Drizin/InterpolatedStrings#1](https://github.com/Drizin/InterpolatedStrings/pull/1)
> 
> Man, you're fast!! 🚀

Naw, just already had it done...

**Comment by Drizin** [comment]
I've published an initial draft.

Most of my efforts were into this new project: https://github.com/Drizin/InterpolatedSql
All string interpolation was moved to this project, and it no longer depends on Dapper.
[InterpolatedSqlBuilder[T]](https://github.com/Drizin/InterpolatedSql/blob/main/src/InterpolatedSql/InterpolatedSqlBuilder%7BT%7D.cs) uses Fluent Builder with Recursive Generics pattern so that all subtypes can have fluent API. And besides that, it's also a little more extensible

I haven't spent much time into DapperQueryBuilder itself - it looks like it's much cleaner now (and all tests are passing), but there's still much room for improvement. I'll be happy to hear your thoughts. I've invited you two to be collaborators (I think it's required to have PR review permissions). 

PR: https://github.com/Drizin/DapperQueryBuilder/pull/62

@terryaney @jehhynes

**Comment by Drizin** [comment]
I haven't tested it yet in any real project (and probably I'll only be able to do it next weekend), 
but in case anyone would like to give it a try I've just published 
[Nuget package 2.0.0-beta1](https://www.nuget.org/packages/Dapper-QueryBuilder/2.0.0-beta1)

**Comment by jehhynes** [comment]
For this rewrite, will it be possible to make the regions fully customizable? It would be super cool if I were able to do something like this:

```
            var q = new QueryBuilder(Connection, $@"
select
	MyInnerTable.*
	/**select*/
from
	(
		select 
			Id
			/**select:somekeysuffix**/
		from 
			SomeTable
			/**from:somekeysuffix**/
		where
			1=1
			/**filters:somekeysuffix**/
	) MyInnerTable
where
	1=1
	/**filters**/
group by
	/**groupby**/
order by
	/**orderby**/

/**myCustomUnion**/
")
               .From($"join OtherTable t on SomeTable.OtherTableId=t.Id", "somekeysuffix")
               .Select($"SomeColumn", "somekeysuffix")
               .Where($"SomeColumn={someValue}", "somekeysuffix")
               .Where($"OtherColumn={otherValue}")
               .GroupBy($"SomeColumn")
               .OrderBy($"SomeColumn")
               .Add("myCustomUnion", $"SomeArbitrarySql");
```

1. The ability to have standard regions (select, from, filters, groupby, orderby)
2. The ability to duplicate the standard regions by using a key suffix (i.e. `select:somekeysuffix`). This would allow for more complex SQL statements that may have inner selects, unions, etc.
3. The ability to have completely custom regions (an `Add(string region, FormattableString sql)` method). 


This does seem like a drastic deviation from the current design but would be so much more useful, and also could be a much cleaner design if done right. It seems like all the regions, both standard and non-standard, could function exactly the same in the back-end. 
- We would only need one collection of Regions. 
- A Region would be described by a Key and a Delimiter string (e.g. "AND", ",", or Environment.NewLine)
  - The delimiter could even be a full formattable string in itself. For instance, we might want to wrap the expression in parentheses: `$" AND ( {0} )"`
- When a FormattableString is added to a region, it adds the Delimiter and then the String
- When using the standard methods with no suffix (Select, From, etc) it would find the Region with a known key (e.g. "select", "from"), and call the `Add()` method for that region
- When using a standard method with a suffix (`Select($"SomeColumn", "somekeysuffix")`) it would find the region with the known key including suffix (e.g. "select:somekeysuffix") and call the `Add()` method for that region
- When using a non-standard region (e.g. "myCustomUnion") it would find the Region with that key and call the `Add()` method for that region.

**Comment by Drizin** [comment]
> For this rewrite, will it be possible to make the regions fully customizable? It would be super cool if I were able to do something like this:

My first thought is that this could be achieved by breaking a query into multiple variables and interpolating one type (like another QueryBuilder) inside the other. Something like this:

```cs
var subquery = new QueryBuilder(null, $@"
 		select 
 			Id
 			/**select**/
 		from 
 			SomeTable
 			/**from**/
 		where
 			1=1
 			/**filters**/
    ")
    .From($"join OtherTable t on SomeTable.OtherTableId=t.Id")
    .Select($"SomeColumn")
    .Where($"SomeColumn={someValue}");

var q = new QueryBuilder(Connection, $@"
    select
        MyInnerTable.*
        /**select*/
    from
        (
            {subquery}
        ) MyInnerTable
    where
        1=1
        /**filters**/
    group by
        /**groupby**/
    order by
        /**orderby**/
    ")
    .Where($"OtherColumn={otherValue}")
    .GroupBy($"SomeColumn")
    .OrderBy($"SomeColumn");
    
FormattableString myCustomUnion = $"SomeArbitrarySql";
// myCustomUnion could also be interpolated directly like "{myCustomUnion}" after the order by above
q += myCustomUnion; 
```

What are your thoughts? I could be biased, but I feel it's simpler and yet flexible. 
Any object implementing [IInterpolatedSql](https://github.com/Drizin/InterpolatedSql/blob/main/src/InterpolatedSql/IInterpolatedSql.cs) can be [just interpolated](https://github.com/Drizin/InterpolatedSql/blob/main/src/InterpolatedSql/InterpolatedSqlParser.cs#L174).

**Comment by jehhynes** [comment]
That is one possibility but with several disadvantages:
1) Less readable
2) What about the parameters?
3) Doesn't solve the root problem which is being able to customize it beyond the standard select/where/from/groupby/orderby clauses.
 - If I have to do gymnastics to use a tool, I'll be less likely to use it. If I have to use multiple query builders and interpolate them manually, I may as well just use a StringBuilder from the start.

**Comment by Drizin** [comment]
I still feel that breaking complex queries into multiple instances (and embedding one inside the other) is easier to read and maintain. But I understand your preference of having a single instance (I assume you're passing that variable across different methods). 

And while I agree that FluentQueryBuilder has some ugly logic that could be refactored, I feel that what you are suggesting is a very specific use case that if added to the core library would cause some confusion among users.

Having said that, I think there's an easy approach here - you can extend SqlBuilder with your own subtype:
```
public class MyQueryBuilder : InterpolatedSql.Dapper.SqlBuilder
{
    public MyQueryBuilder(IDbConnection connection, FormattableString query) : base(connection, query) {}

    public MyQueryBuilder AppendCustomObject(FormattableString value, string key, string separator, string prefixBeforeFirstItem = "")
    {
        base.ObjectBag ??= new System.Collections.Generic.Dictionary<string, object>();
        if (!ObjectBag.ContainsKey(key))
            ObjectBag[key] = new InterpolatedSqlBuilder();
        InterpolatedSqlBuilder builder = (InterpolatedSqlBuilder)base.ObjectBag[key];
        builder.AppendRaw(builder.Length == 0 ? prefixBeforeFirstItem : separator);
        builder.Append(value);
        return this;
    }
    public MyQueryBuilder From(FormattableString value, string key) => AppendCustomObject(value, "from:" + key, "\nINNER JOIN ");
    public MyQueryBuilder Select(FormattableString value, string key) => AppendCustomObject(value, "select:" + key, ", ", ", ");
    public MyQueryBuilder Where(FormattableString value, string key) => AppendCustomObject(value, "filters:" + key, " AND ");
    public MyQueryBuilder Where(FormattableString value) => AppendCustomObject(value, "filters", " AND ");
    public MyQueryBuilder GroupBy(FormattableString value) => AppendCustomObject(value, "groupby", " AND ");
    public MyQueryBuilder OrderBy(FormattableString value) => AppendCustomObject(value, "orderby", " AND ");
    public MyQueryBuilder Add(string key, FormattableString value) => AppendCustomObject(value, key, "");
    public void ReplaceCustomObjects()
    {
        foreach(var key in base.ObjectBag!.Keys) 
            this.Replace("/**" + key + "**/", (InterpolatedSqlBuilder)base.ObjectBag![key]);
    }
}
```

```
string someValue = "hope it";
string otherValue = "works";

var q = new MyQueryBuilder(cn, $$"""
    select
        MyInnerTable.*
        /**select*/
    from
        (
            select 
                Id
                /**select:somekeysuffix**/
            from 
                SomeTable
                /**from:somekeysuffix**/
            where
                1=1
                /**filters:somekeysuffix**/
        ) MyInnerTable
    where
        1=1
        /**filters**/
    group by
        /**groupby**/
    order by
        /**orderby**/

    /**myCustomUnion**/
    """)
        .From($"join OtherTable t on SomeTable.OtherTableId=t.Id", "somekeysuffix")
        .Select($"SomeColumn", "somekeysuffix")
        .Where($"SomeColumn={someValue}", "somekeysuffix")
        .Where($"OtherColumn={otherValue}")
        .GroupBy($"SomeColumn")
        .OrderBy($"SomeColumn")
        .Add("myCustomUnion", $"SomeArbitrarySql");
;
q.ReplaceCustomObjects();

System.Diagnostics.Debug.WriteLine(q.Sql);
System.Diagnostics.Debug.WriteLine(q.DapperParameters["p0"]);
System.Diagnostics.Debug.WriteLine(q.DapperParameters["p1"]);
```

As you can see, there's this ObjectBag (Dictionary) which can be used for customizations.
(Yes, parameters are merged - not only when Appending but also when Replacing).

Note that I'm using `InterpolatedSql.Dapper.SqlBuilder` and no`InterpolatedSql.Dapper.FluentQueryBuilder.FluentQueryBuilder`, so you lose some of the out-of-the-box features, but you can implement on your own.
If you look at the [new codebase](https://github.com/Drizin/InterpolatedSql) you'll see that some classes (like `InterpolatedSqlBuilder`) are using the Fluent Builder with Generics Inheritance pattern, which means that subtypes still get the whole Fluent API (subtype passes its generic type `T` to the parent class, then all methods will always return subtype `T`). I should probably add this same behavior to `SqlBuilder` and `FluentQueryBuilder` so that subtypes can still benefit from the Fluent API from parent types.
(In other words, in my example above if you used `.Append()` you would get back an `SqlBuilder`, instead of the custom `MyQueryBuilder`. 

To sum up, there is still much room for improvement, but I'd rather avoid use cases that are very specific in favor of a more extensible library.

Hope it makes sense.

**Comment by jehhynes** [comment]
@Drizin Thanks for considering the possibility and offering a workaround.

**Comment by Drizin** [comment]
Major rewrite / release candidate (2.0.0-beta3): https://github.com/Drizin/InterpolatedSql 

@jehhynes / @terryaney I've refactored a bunch of things, there's a lot of generics now, and keeping legacy compatibility was a little tricky - but hopefully the code should be easy to follow.

I'm open to suggestions (in the other project please - I'll close this issue and probably will soon archive this repo)

@jehhynes - extensibility example [here](https://github.com/Drizin/InterpolatedSql/blob/main/src/InterpolatedSql.Dapper.Tests/ExtensibilityTests.cs)

**Comment by jehhynes** [comment]
@Drizin Very nice, I love the extensibility example. Well done.
